File: client/GameRelay.Core/Protocol.cs
using System.Buffers.Binary;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace GameRelay.Core;

/// <summary>
/// JSON envelope for every control-channel frame. Field names must match
/// the Go relay server exactly.
/// </summary>
public sealed class ControlMessage
{
    public const int ProtocolVersion = 1;

    [JsonPropertyName("type")] public string Type { get; set; } = "";

    [JsonPropertyName("nonce")] public string? Nonce { get; set; }
    [JsonPropertyName("mac")] public string? Mac { get; set; }
    [JsonPropertyName("role")] public string? Role { get; set; }
    [JsonPropertyName("conn_id")] public string? ConnId { get; set; }
    [JsonPropertyName("version")] public int Version { get; set; }

    [JsonPropertyName("udp_port")] public int UdpPort { get; set; }

    [JsonPropertyName("tunnel_id")] public string? TunnelId { get; set; }
    [JsonPropertyName("proto")] public string? Proto { get; set; }
    [JsonPropertyName("public_port")] public int PublicPort { get; set; }
    [JsonPropertyName("reason")] public string? Reason { get; set; }

    [JsonPropertyName("remote_addr")] public string? RemoteAddr { get; set; }

    [JsonPropertyName("ts")] public long Ts { get; set; }
}

/// <summary>Control message type constants (mirror of the Go side).</summary>
public static class MessageTypes
{
    public const string Challenge = "challenge";
    public const string Auth = "auth";
    public const string AuthOk = "auth_ok";
    public const string Error = "error";
    public const string OpenTunnel = "open_tunnel";
    public const string CloseTunnel = "close_tunnel";
    public const string TunnelOk = "tunnel_ok";
    public const string TunnelFail = "tunnel_fail";
    public const string ConnRequest = "conn_request";
    public const string Ping = "ping";
    public const string Pong = "pong";
}

/// <summary>
/// Length-prefixed JSON framing: 4-byte big-endian length + UTF-8 JSON.
/// </summary>
public static class Framing
{
    public const int MaxFrameSize = 64 * 1024;

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    public static async Task WriteAsync(Stream stream, ControlMessage msg, CancellationToken ct)
    {
        byte[] body = JsonSerializer.SerializeToUtf8Bytes(msg, JsonOpts);
        if (body.Length > MaxFrameSize)
            throw new InvalidOperationException($"frame too large: {body.Length}");
        byte[] buf = new byte[4 + body.Length];
        BinaryPrimitives.WriteUInt32BigEndian(buf, (uint)body.Length);
        body.CopyTo(buf, 4);
        await stream.WriteAsync(buf, ct).ConfigureAwait(false);
    }

    public static async Task<ControlMessage> ReadAsync(Stream stream, CancellationToken ct)
    {
        byte[] head = new byte[4];
        await stream.ReadExactlyAsync(head, ct).ConfigureAwait(false);
        uint len = BinaryPrimitives.ReadUInt32BigEndian(head);
        if (len == 0 || len > MaxFrameSize)
            throw new InvalidDataException($"invalid frame size {len}");
        byte[] body = new byte[len];
        await stream.ReadExactlyAsync(body, ct).ConfigureAwait(false);
        return JsonSerializer.Deserialize<ControlMessage>(body, JsonOpts)
               ?? throw new InvalidDataException("null frame");
    }
}

/// <summary>UDP carrier datagram constants (mirror of the Go side).</summary>
public static class UdpProtocol
{
    public const byte Magic = 0xC7;
    public const byte TypeAuth = 0x01;
    public const byte TypeAuthOk = 0x02;
    public const byte TypeKeepAlive = 0x03;
    public const byte TypeData = 0x04;
    public const int HeaderLen = 1 + 1 + 4 + 2; // magic + type + session + port
}

File: server/protocol.go
package main

// Wire protocol shared between the relay server and the Windows client.
//
// Control / data channel (TCP):
//   Every frame is a 4-byte big-endian length followed by a JSON object.
//   The connection starts with a challenge/response HMAC handshake, after
//   which the connection either becomes the control channel (role=control)
//   or a raw data pipe for one visitor connection (role=data).
//
// UDP carrier (single UDP socket on the client side):
//   Every datagram starts with [magic 0xC7][type byte].
//   DATA frames add [session id uint32 BE][public port uint16 BE][payload].

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"time"
)

const (
	// ProtocolVersion is bumped on any incompatible wire change.
	ProtocolVersion = 1

	// MaxFrameSize bounds a single control frame. Control messages are
	// small; anything larger indicates a broken or malicious peer.
	MaxFrameSize = 64 * 1024

	// UDP carrier datagram types.
	udpMagic        byte = 0xC7
	udpTypeAuth     byte = 0x01
	udpTypeAuthOK   byte = 0x02
	udpTypeKeepAliv byte = 0x03
	udpTypeData     byte = 0x04

	// udpHeaderLen is the DATA frame header: magic+type+session+port.
	udpHeaderLen = 1 + 1 + 4 + 2

	// udpAuthWindow is how far a carrier AUTH timestamp may drift.
	udpAuthWindow = 5 * time.Minute
)

// Message is the single JSON envelope for every control-channel frame.
// Only the fields relevant to each Type are populated.
type Message struct {
	Type string `json:"type"`

	// challenge / auth
	Nonce   string `json:"nonce,omitempty"`
	MAC     string `json:"mac,omitempty"`
	Role    string `json:"role,omitempty"`    // "control" or "data"
	ConnID  string `json:"conn_id,omitempty"` // for role=data
	Version int    `json:"version,omitempty"`

	// auth_ok
	UDPPort int `json:"udp_port,omitempty"`

	// tunnels
	TunnelID   string `json:"tunnel_id,omitempty"`
	Proto      string `json:"proto,omitempty"` // "tcp" or "udp"
	PublicPort int    `json:"public_port,omitempty"`
	Reason     string `json:"reason,omitempty"`

	// conn_request
	RemoteAddr string `json:"remote_addr,omitempty"`

	// ping / pong
	TS int64 `json:"ts,omitempty"`
}

// Control message types.
const (
	MsgChallenge   = "challenge"
	MsgAuth        = "auth"
	MsgAuthOK      = "auth_ok"
	MsgError       = "error"
	MsgOpenTunnel  = "open_tunnel"
	MsgCloseTunnel = "close_tunnel"
	MsgTunnelOK    = "tunnel_ok"
	MsgTunnelFail  = "tunnel_fail"
	MsgConnRequest = "conn_request"
	MsgPing        = "ping"
	MsgPong        = "pong"
)

// WriteFrame serialises msg and writes one length-prefixed frame.
func WriteFrame(w io.Writer, msg *Message) error {
	body, err := json.Marshal(msg)
	if err != nil {
		return err
	}
	if len(body) > MaxFrameSize {
		return fmt.Errorf("frame too large: %d bytes", len(body))
	}
	buf := make([]byte, 4+len(body))
	binary.BigEndian.PutUint32(buf[:4], uint32(len(body)))
	copy(buf[4:], body)
	_, err = w.Write(buf)
	return err
}

// ReadFrame reads one length-prefixed JSON frame.
func ReadFrame(r io.Reader) (*Message, error) {
	var head [4]byte
	if _, err := io.ReadFull(r, head[:]); err != nil {
		return nil, err
	}
	n := binary.BigEndian.Uint32(head[:])
	if n == 0 || n > MaxFrameSize {
		return nil, fmt.Errorf("invalid frame size %d", n)
	}
	body := make([]byte, n)
	if _, err := io.ReadFull(r, body); err != nil {
		return nil, err
	}
	var msg Message
	if err := json.Unmarshal(body, &msg); err != nil {
		return nil, err
	}
	return &msg, nil
}

// setNoDelay disables Nagle's algorithm where possible to keep latency low.
func setNoDelay(c net.Conn) {
	if tc, ok := c.(*net.TCPConn); ok {
		_ = tc.SetNoDelay(true)
	}
}
